### ISSUE TITLE
Fill pattern scaling based on pixel ratios

### ISSUE THREAD

**Issue by mike-000** [openlayers/openlayers#15299]
New issue arising from https://github.com/openlayers/openlayers/pull/15279#issuecomment-1793674443 as it is apparent neither canvas or WebGL fill patterns match Mapbox, or one another.

> we're trying to aim for supporting Mapbox Style features, and the spec does not offer any option for scaling fill patterns.

Scaling is not an option in Mapbox, however patterns are scaled based on pixel ratios because the sprite used is based on pixel ratio.

When using Openlayers canvas https://codesandbox.io/s/vector-esri-forked-3ly3ry?file=/main.js is programmed to match the pixel ratio (any pixel ratio), so the rendered size of the patterns in css pixels is constant.

But convert that to WebGL and force a pixel ratio of 2 in WebGL and the size doubles https://jsfiddle.net/j4ozv3x9/

Currently for patterns in a Mapbox style the relative rendered sizes in css pixels (based on the size of wave patterns in https://openlayers.org/en/main/examples/mapbox-vector-layer.html and
https://openlayers.org/en/main/examples/mapbox-layer.html, and inferred from the behaviour above) would be

```
Pixel   Sprite   Mapbox   OpenLayers  OpenLayers
ratio   size              canvas      WebGL

1       1        1/1      1/1         1/1

2       2        2/2      2/2         2/1

3       2        2/2      2/3         2/1
```

Mapbox scales based on predefined sprite pixel ratios, while OpenLayers canvas will scale based on any pixel ratio even when dealing with predefined Mapbox sprites (ol-mapbox-style simply uses the closest sprite) and currently WebGL always renders to the original css pixel size.

To better support Mapbox styles both canvas and WebGL should be able to scale to the same sprite pixel ratios as Mapbox.

**Comment by jahow** [comment]
So basically:
* draw patterns with a constant screen pixel size (e.g. twice smaller if DPI is 2), and
* somehow allow using the pixel ratio in the style expressions to switch which spritesheet to use?

**Comment by mike-000** [comment]
Yes, both Mapbox and ol-mapbox-style choose the best sprite for the device/map pixel ratio.
Both scale icons to the sprite pixel ratio (ol-mapbox-style reduces the icon-scale if necessary).
Mapbox appears to do the same for patterns, while ol-mapbox-style leaves pattern scaling to the canvas renderer, which uses the map pixel ratio.  It is important that pattern sizes divisible by 512 align at tile edges.

**Comment by ahocevar** [comment]
I think once we introduce sprite sheets, some of the open questions here can be answered more easily. Then sprite sheets could be registered for a specific pixel ratio (e.g. 1 or 2), and then everything taken from that style sheet (icons, patterns) will be scaled to match the registered pixel ratio.

**Comment by jahow** [comment]
I added a `fill-pattern-scale` option in #15317 because I noticed that a pattern that is 50x50 will be rendered seamlessly in Maplibre but was not lining up properly at the tiles' edges in the OL webgl renderer. So at least now we have more flexibility to handle different cases like the ones you mentioned. A stylesheet system like the one from Maplibre can be introduced later on to hide all this complexity to the user and generate the settings itself.
